archive/tar.go

```diff
@@ -247,18 +256,20 @@ func Untar(r io.Reader, dest string) error {
 			if err := os.MkdirAll(path, os.ModePerm); err != nil {
 				return err
 			}
+			dirsFound[path] = true
+
 		case tar.TypeReg, tar.TypeRegA:
-			_, err := os.Stat(filepath.Dir(path))
-			if os.IsNotExist(err) {
-				if err := os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
-					return err
+			dirPath := filepath.Dir(path)
+			if !dirsFound[dirPath] {
+				if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+					if err := os.MkdirAll(dirPath, os.ModePerm); err != nil {
```

Comment:
I believe this PR has changed the behavior that governs the directory permissions in this particular case. If there is no header for a directory in the archive, the umask should be applied. Because we have unset the umask, directories that fall into this case will all be created with perm `0777`.

Reply:
Thanks for spotting that! I'll fix that and most probably add a test case